proposals/unsafe-evolution.md

```diff
@@ -0,0 +1,138 @@
+        ref int intRef = Unsafe.AsRef(ptr); // Conversion of memory not managed by the runtime to a `ref`. This is unsafe.
+    }
+}
+```
+
+## Motivation
+
+Background for this feature can also be found in https://github.com/dotnet/designs/pull/330, which tracks the broader ecosystem changes that will be needed as part of this proposal. For C#
+specifically, we want to make sure that memory unsafety is properly tracked by the language; today, it can be difficult to look at a program hollistically and understand all locations where
+memory unsafety occurs. This is because various helpers such as the `System.Runtime.CompilerServices.Unsafe`, `System.Runtime.InteropServices.Marshal`, and others do not express that they
+violate memory safety and need special consideration. Methods that then use these helpers aren't immediately obvious, and when auditing code for memory safety issues (either ahead of time
+when doing review, or when trying to determine the cause of a vulnerability that is being reported) it can be difficult to pinpoint the locations that could be contributing to issues.
+
+Historically, `unsafe` in C# has referred to a specific memory-safety hole: the existence of pointer types. The moment that a pointer type is no longer involved, C# is perfectly happy to let
+memory unsafety lie latent in code. It is this issue that we are looking to address with this evolution of `unsafe` in C# and the .NET ecosystem, labeling areas where memory unsafety could
+potentially being occuring, making it easier for reviewers and auditors to understand the boundaries of potential memory unsafety in a program. Importantly, this means that we will be _changing_
+the meaning of `unsafe`, not just augmenting it. The existence of a pointer is not itself unsafe; the unsafe action is dereferencing the pointer. This extends further to types themselves;
+types cannot be inherently unsafe. It is only the action of using a type that could be unsafe, not the existence of that type.
+
+In order for this information to flow through the system, we therefore need to have a way to mark methods themselves as `unsafe`. Today, `unsafe` as a method modifier has no external impact,
+it only allows pointers to be used in the body of the member. Going forward, `unsafe` as a modifier will actually publicly change the meaning of the member; it will indicate that the member
+has memory safety concerns and any usages must be manually validated by the programmer using the member.
+
+This is a potentially large breaking change for particular segments of the C# user base. Our hope is that, for many of our users, this is effectively transparent, and updating to the new rules
+will be seemless. However, given that some large API surfaces like large parts of reflection may need to be marked `unsafe`, we do think it likely that there will need to be a decent on-ramp to
+the new rules to avoid entirely bifurcating the ecosystem.
+
+## Detailed Design
```

Comment:
Iterators currently escape out of `unsafe` contexts (i.e., iterator blocks implicitly introduce a 'safe' context) and it's not allowed to use `yield`/`await` inside `unsafe` contexts. See also https://github.com/dotnet/csharplang/blob/main/proposals/csharp-13.0/ref-unsafe-in-iterators-async.md. I guess some of that could be changed by this proposal?